### PR TITLE
ICMSLST-1035 Add wildcard filter function for searching applications. 

### PR DESCRIPTION
**Pull request by MattHolmes123**
Other changes:

Add several indexes to core application tables.

### CODE REVIEW

File: web/utils/search.py [hunk 1]
```diff
@@ -792,6 +791,37 @@ def _apply_export_application_filter(
     return model
 
 
+def _apply_wildcard_filter(
+    model: "QuerySet[Model]", field: str, search_pattern: str
+) -> "QuerySet[Model]":
+    """Strings with `%` are converted into django orm code using one of several methods.
```
Comment (oskusalerma) [inline web/utils/search.py:797]: s/orm/ORM/

File: web/utils/search.py [hunk 2]
```diff
@@ -792,6 +791,37 @@ def _apply_export_application_filter(
...
+    :param field: The name of the field to search on
+    :param search_pattern: the user supplied search pattern
+    """
+
+    if search_pattern.strip() == "%":
+        return model
+
+    # The default search unless changed below
+    search_regex = search_pattern.replace("%", ".*")
+    search = {f"{field}__iregex": f"^{search_regex}"}
```
Comment (oskusalerma) [inline web/utils/search.py:809]: if we're doing regex search, do we need to escape regex characters in user supplied string?

**Review by oskusalerma — CHANGES_REQUESTED** [review]

File: web/utils/search.py (hunk 2 above)
Comment (MattHolmes123) [reply web/utils/search.py:809]: Probably, I haven't really thought about it yet

Comment (MattHolmes123) [reply web/utils/search.py:809]: Added a subtask for this.

File: web/utils/search.py [hunk 3]
```diff
@@ -270,6 +270,36 @@ def get_search_results_spreadsheet(case_type: str, results: SearchResults) -> by
...
+    :param field: The name of the field to search on
+    :param search_pattern: the user supplied search pattern
+    """
+
+    if search_pattern.strip() == "%":
+        return {}
+
+    # The default search unless changed below
+    search_regex = search_pattern.replace("%", ".*")
+    search = {f"{field}__iregex": f"^{search_regex}"}
```
Comment (mikeylockwood) [inline web/utils/search.py:287]: Does the regex need to be escaped before creating a regex search?

```
search_regex = re.escape(search_pattern).replace("//%", ".*")
"""

Comment (mikeylockwood) [reply web/utils/search.py:287]: Maybe that won't work with  `istartswith` and `iendswith` below

Comment (MattHolmes123) [reply web/utils/search.py:287]: I have a story to escape the regex later.
re.escape isn't necessarily the correct thing to do.

**Review by mikeylockwood — APPROVED** [review]

**Review by kevinetienne — APPROVED** [review]

**Pull request merged**